[sprint_8-solution] refactor and add new logic:

PascalABCD opened the pull request:
- add Duration and StartTime
- fix tests
- check overlapping Tasks by startTime
- create get tasks by Priority. Priority calculates by start time

PascalABCD committed:
[sprint_8-solution] refactor and add new logic:

- add Duration and StartTime
- fix tests
- check overlapping Tasks by startTime
- create get tasks by Priority. Priority calculates by start time

PascalABCD committed:
[sprint_8-solution] fix tests

PascalABCD committed:
[sprint_8-solution] fix tests

PascalABCD committed:
[sprint_8-solution] fix tests

irek-valiev commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -29,17 +29,13 @@ public void setId(int id) {
-        for (Task task : tasks.values()) {
-            hm.remove(task.getId());
-        }
+        tasks.values().forEach(task -> hm.remove(task.getId()));
```
нужно еще из prioritizedTasks удалять

irek-valiev commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -56,17 +52,31 @@ public Task getTaskById(int taskId) {
         if (tasks.containsKey(taskId)) {
             tasks.replace(taskId, newTask);
+            if (newTask.getStartTime() != null) {
+                prioritizedTasks.add(newTask);
```
сначала из prioritizedTasks удаляется старая версия, потом обновляется tasks, потом prioritizedTasks. 2 и 3 шаги можно поменять местами

irek-valiev commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -87,22 +98,19 @@ public void removeTask(Task task) {
-        }
-
-        for (Epic epic : epics.values()) {
+        subtasks.values().forEach(subtask -> hm.remove(subtask.getId()));
```
нужно еще из prioritizedTasks удалять

irek-valiev commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -87,22 +98,19 @@ public void removeTask(Task task) {
             updateEpicStatus(epic);
-        }
+            updateEpicTimeFields(epic);
+            prioritizedTasks.remove(epic);
```
эпиков там быть не должно, они в prioritizedTasks представлены своими подзадачами

irek-valiev commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -145,6 +163,8 @@ public Subtask updateSubtask(Subtask newSubtask) {
                 subtasks.replace(subtaskId, newSubtask);
                 updateEpicStatus(epic);
+                updateEpicTimeFields(epic);
+                prioritizedTasks.add(newSubtask);
```
сначала из prioritizedTasks удаляется старая версия, потом обновляется subtasks, потом prioritizedTasks. 2 и 3 шаги можно поменять местами

irek-valiev commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -208,6 +229,7 @@ public void createEpic(Epic epic) {
         epic.setId(epicId);
 
         epics.put(epic.getId(), epic);
+        prioritizedTasks.add(epic);
```
эпики не хранятся в prioritizedTasks, они представлены там своими подзадачами

irek-valiev commented on src/model/Epic.java:
```diff
@@ -1,16 +1,50 @@
-    private ArrayList<Subtask> subtasksList = new ArrayList<>();
+    protected ArrayList<Subtask> subtasksList = new ArrayList<>();
+    protected LocalDateTime endTime;
 
```
у Эпика при создании все атрибуты времени должны быть null, потому что они появятся только с первой подзадачей, у которой заполнены startTime и duration. Т.е. эпик может иметь сколько угодно подзадач, но если ни у одной из них не заполнено время, то у него время должно остаться null.

irek-valiev commented on src/model/Epic.java:
```diff
@@ -1,16 +1,50 @@
+    public void calculateEpicDuration() {
+        duration = subtasksList.stream()
+                .map(Subtask::getDuration)
+                .filter(subtaskDuration -> subtaskDuration != null)
```
замени на filter(Objects::nonNull)

irek-valiev commented on src/model/Epic.java:
```diff
@@ -1,16 +1,50 @@
+    }
+
+    public void calculateEpicEndTime() {
+        LocalDateTime maxSubtaskEndTime = subtasksList.stream()
```
вот тут сделай .map(Subtask::getEndTime)

irek-valiev commented on src/model/Epic.java:
```diff
@@ -1,16 +1,50 @@
+
+    public void calculateEpicEndTime() {
+        LocalDateTime maxSubtaskEndTime = subtasksList.stream()
+                .filter(subtask -> subtask.getEndTime() != null)
```
замени на filter(Objects::nonNull)

irek-valiev commented on src/model/Epic.java:
```diff
@@ -1,16 +1,50 @@
+    }
+
+    public void calculateEpicStartTime() {
+        LocalDateTime minSubtaskstartTime = subtasksList.stream()
```
здесь тоже последовательность getStartTime->nonNull->Comparator.comparing

irek-valiev commented on src/model/Subtask.java:
```diff
@@ -1,17 +1,19 @@
 public class Subtask extends Task {
-    private int epicId;
+    protected int epicId;
 
```
переопределить еще родительский конструктор без полей Duration duration, LocalDateTime startTime

irek-valiev commented on src/model/Task.java:
```diff
@@ -1,26 +1,37 @@
+    protected Status status;
+    protected Duration duration;
+    protected LocalDateTime startTime;
+    protected LocalDateTime endTime;
```
эта переменная  понадобится только в классе  Epic

irek-valiev commented on src/model/Task.java:
```diff
@@ -1,26 +1,37 @@
+    protected LocalDateTime startTime;
+    protected LocalDateTime endTime;
+
+    public Task(String name, String description, Status status, Duration duration, LocalDateTime startTime) {
```
пользователь при создании может не знать когда приступит к задаче и сколько на нее понадобится времени, поэтому нужен еще конструктор без этих полей

irek-valiev commented on src/model/Task.java:
```diff
@@ -53,6 +64,21 @@ public void setStatus(Status status) {
+    }
+
+    public LocalDateTime getEndTime() {
+        return startTime.plus(duration);
```
нужно добавить проверку на null для startTime, поскольку пользователь может создавать задачи не указывая сразу их время, он может пока их не знать

irek-valiev replied:
Что-то проверка на null не появилась

PascalABCD committed:
[sprint_8-solution] review fixes

irek-valiev commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -125,26 +139,39 @@ public void createSubtask(Subtask subtask) {
         int subtaskId = newSubtask.getId();
         if (subtasks.containsKey(subtaskId)) {
             Subtask oldSubtask = subtasks.get(subtaskId);
+            prioritizedTasks.remove(oldSubtask);
```
это действие тоже сделай внутри  if (oldSubtask.getEpicId() == newSubtask.getEpicId())

PascalABCD committed:
[sprint_8-solution] review fixes:

- add null check for startTime in get getEndTime();
- move down remove(oldSubtask) in updateSubtask
